### DISCUSSION TITLE
Understanding sessions

### DISCUSSION THREAD

**Discussion by amit-chandak-unskript** [jupyter/enterprise_gateway#1005] [Q&A]
We are using jupyterlab with enterprise gateway. The way we render jupyter notebook is in an iframe, by constructing the url for the ipynb file we want to show, for eg, something like **https://<jupyter server url>/lab/tree/<ipynb file>**. This worked fine, however what we saw that bunch of jupyter sessions  kept lingering around even though we moved away from the page. Probably, we were not doing **Close and Shutdown** kind of close on the files opened in that iframe. Eventually, the issue we hit was after few of these sessions were present, going to a file **foo.ipynb** landed up on file **bar.ipynb**. I havent fully understand how that could happen, but at this point, i wanted to understand sessions.

* Is session per notebook or the session be shared for multiple notebooks?
* If its per notebook, then i can do a close and shutdown to close the session, but with enterprise gateway, it will shutdown the kernel too.
* AFAIK, with EG, multiple notebooks can share the same kernel, is it a valid use case or the main use case is to have per notebook kernel?
* Any idea how these sessions are initiated? I tried looking at the developer tool, but i dont see any POST on session api.
* Any pointer on understanding sessions? 

Thanks for the help. @kevin-bates

**Comment by kevin-bates** [comment]
Hi @amit-chandak-unskript,

This sounds more like a question for the Lab folks.  When the Lab (or Notebook) server is configured to use EG, the server essentially redirects kernel management and websocket requests to the EG server running in a headless mode.  The sessions used (and I believe those you see leaked) are in the lab/notebook server since EG doesn't use these sessions  (EG has a _kernel session_ object for managing state for HA/DR purposes, but that is different).

I'm not intimately familiar with how the frontends manage kernels but can try to provide what I know relative to your questions.

> Is session per notebook or the session be shared for multiple notebooks?

A session is created per notebook since it's via the session that kernels are initiated.

> If its per notebook, then i can do a close and shutdown to close the session, but with enterprise gateway, it will shutdown the kernel too.

This is because shutdown of the notebook is supposed to stop the kernel and the notebook server is just relaying that kernel management aspect of things to EG.

> AFAIK, with EG, multiple notebooks can share the same kernel, is it a valid use case or the main use case is to have per notebook kernel?

Yes.  The same goes for lab. You can change a kernel of an existing notebook to another, currently running, kernel.  EG is merely a "kernel as a service" model - so hitting the appropriate endpoint with the appropriate kernel_id will access the same kernel.

> Any idea how these sessions are initiated? I tried looking at the developer tool, but i dont see any POST on session API.

For EG, kernels are started via a POST to /api/kernels, whereas the front-end issues a POST to /api/sessions.  That latter case, under the covers when EG is configured, then issues the POST to /api/kernels.

**Reply by amit-chandak-unskript** [reply]
Thanks a lot @kevin-bates , this aligns with what i have understood so far, looking at the code!!!
